### ISSUE TITLE
macOS: Can't solve "Mouse: not useable (blocked)"

### ISSUE THREAD

**Issue by Gudzo** [RaiMan/SikuliX1#429]
I was using Sikuli API with Java 8 and Maven on a macOS, Windows and Linux Mint, everything was working fine with macOS Mojave. Hadn't touched the project in a while, upgraded to Catalina 10.15.7,  got back to it this week, saw that there is a new api and and a new IDE 2.0.5, downloaded it, imported maven dependency 2.0.5 api in Maven config. Started the IDE and got the error <img width="1439" alt="ide" src="https://user-images.githubusercontent.com/14196338/110619855-9b924a80-8198-11eb-8fb3-12e128604f76.png">,
 tried to run the test from Intelij Idea, got the same error 
<img width="910" alt="idea error" src="https://user-images.githubusercontent.com/14196338/110619965-c2508100-8198-11eb-9b64-e175361109ba.png">. 
Went to Security & Privacy and I am seeing everything that can be enabled in Accessibility
<img width="668" alt="Accessibility 1" src="https://user-images.githubusercontent.com/14196338/110620216-0d6a9400-8199-11eb-85d7-3c0aa21bdafc.png">
<img width="671" alt="Accessibility 2" src="https://user-images.githubusercontent.com/14196338/110620230-10fe1b00-8199-11eb-9190-f95b89d402ba.png">
<img width="672" alt="Accessibility 3" src="https://user-images.githubusercontent.com/14196338/110620234-122f4800-8199-11eb-99b9-804b4c418558.png"> and Screen Recording
<img width="670" alt="screen 1" src="https://user-images.githubusercontent.com/14196338/110620264-1bb8b000-8199-11eb-884c-68771038dbaf.png">
<img width="675" alt="screen 2" src="https://user-images.githubusercontent.com/14196338/110620275-1fe4cd80-8199-11eb-814f-3f4473e600d3.png"> is actually enabled. 

Has anything similar happened to anybody before?

**Comment by RaiMan** [comment]
Not sure, wether there are differences between macOS 10.15.7 and macOS 11.2.2 (what I use for testing). 

I just tested again on my Mac and found, that everything is as expected and written in the mentioned wiki page on GitHub.

- download the IDE
- in a Terminal (the Mac app!) use `java .jar sikulix...jar` (preferably Java 8 or 11)
- you should get the system popup (click goto ...) and a bit later the SikuliX popup
- if the Terminal entry is already in the Prefs, but not activated, you only get the SikuliX popup
- after having activated the Terminal entries in the Prefs, everything works as expected

Same goes for the case, where you run a project from inside IDEA, the relies on a SikuliX 2.0.5 jar.
In this case, you get the exception.

**Comment by RaiMan** [comment]
In doubt:
- delete (-) the respective entries in the Prefs
- stop the respective apps 
- restart the system
- start all over again

**Comment by Gudzo** [comment]
So I removed everything from Accessibility and Screen Recording, restarted the machine, started the ide jar using terminal app(I was using iterm before) gave me a prompt to enable Accessibility but nothing for screen recording, it's just blank now. I googled for issues on Catalina regarding screen recording and stumbled upon this https://apple.stackexchange.com/questions/384230/how-do-i-reset-screen-recording-permission-on-macos-catalina, I restarted it using tccutil reset ScreenCapture, deleted all my allowed apps in Screen Recording, again restarted the machine, started the ide jar, still getting same prompt and error message on ide "[error] Mouse: not useable (blocked)".

Basically without any ideas atm.

**Comment by RaiMan** [comment]
The popup about accessibility you get on IDE start.
The screen recording popup only comes up later, when you try to take a screenshot in the IDE. If not allowed, the screen will only show your screen background (no apps, no windows).

So first make Accessibility ready for the Terminal app (in doubt remove before starting IDE)

No need to delete any other entries besides Terminal.

Then after getting around the mouse-blocked-problem deal with the screenshot problem.

When you got that, then deal with IDEA and iTerm

**Comment by pp082** [comment]
Can you validate if the accessibility accesses were actually added, via terminal ?

sqlite3  /Library/Application\ Support/com.apple.TCC/TCC.db 'SELECT * from access' | grep kTCCServiceAccessibility

In case of access problems, you must allow the terminal access to the disk and run again.

- To recognize the position of the authorizing flag, you must know the structure of the table
sudo sqlite3 /Library/Application\ Support/com.apple.TCC/TCC.db 'pragma table_info(access) ;'



When finished you can remove the disk permissions to the terminal :)

**Comment by RaiMan** [comment]
@pp082 
Thanks for your input.

Just checked your approach:
- the sqlite command does not list Terminal, though it is shown in the SysPrefs and everything works well with SikuliX
- running Java 8, then not Terminal is the needed entry, but java (with Java 11 it is Terminal)

When I run SikuliX with Java 8 in the Terminal, I get the known popups. When I now activate the java entry in Accessibility, it is not listed with the sqlite command. When changes in the SysPrefs are made, the IDE must be restarted. Wether Terminal is activated or not, does not matter for Java 8.

**I now found out:**
If you add an app or an executable using the + in the SysPreps, then the entry shows up immediately, when using the sqlite command.

So I guess, that the best approach is, to actively add the respective app or executable  using the + in the SysPreps, because any changes done using - or + button in SysPrefs instantly show up using the sqlite command.

So the steps are the following:
- **in any case stop the Terminal app completely (not only windows!)**

- **when you want to run the SikuliX IDE from a Terminal using Java 8**
  -  first in SysPrefs remove all entries with the name **java** (using - button)
  - then use the + button to add java again: in the file dialog navigate to the respective java executable (..../bin/java)
  - do this for Accessibility and Screenrecording
  - start a Terminal window
  - run the Java command to start the SikuliX IDE

- **when you want to run the SikuliX IDE from a Terminal using Java 11+**
  -  first in SysPrefs remove the entry with the name **Terminal** (using - button)
  - then use the + button to add Terminal again: in the file dialog navigate to the Terminal.app
  - do this for Accessibility and Screenrecording
  - start a Terminal window
  - run the Java command to start the SikuliX IDE

IMHO it is neither necessary to remove any other entries nor to restart the system.

Feedback appreciated.

**Comment by pp082** [comment]
@RaiMan thanks, in my case, understanding that the permission was missing, I did the following:

**Add permission via terminal**
`sudo sqlite3 "/Library/Application Support/com.apple.TCC/TCC.db" 'insertar en valores de acceso ("kTCCServiceAccessibility", "com.apple.terminal", 1, 2, 4, 1", "", "", "", "UNUSED", "", 0,1614632925)'`

But it didn't work because the file is protected by SIP. 

To disable SIP:
Turn off your Mac (Apple > Shut Down).
Hold down Command-R and press the Power button. ...
Wait for OS X to boot into the OS X Utilities window.
Choose Utilities > Terminal.
Enter csrutil disable.
Enter reboot.

_At the login again you have several options, but it is always advisable to validate that the accesses were added correctly. **Remember to re-enable SIP at the time of completion**._

**In my case:**

- **Add the permission via terminal**

`sqlite3 "/Library/Application Support/com.apple.TCC/TCC.db" 'insertar en valores de acceso ("kTCCServiceAccessibility", "com.apple.terminal", 0, 1, 1, 1", "", "", "", "UNUSED", "", 0,1614632925)'` 

- **Test the app and check again via terminal if it was added.**

`sqlite3 /Library/Application\ Support/com.apple.TCC/TCC.db 'SELECT * from access' | grep kTCCServiceAccessibility`

**Doing it in a conventional way, that is, eliminating and adding the options again is totally valid, you only have to worry about checking that they were correctly entered.**


I investigated for about 3 weeks the reason for the message and why it did not work via the terminal, now it works perfectly!. 

**_Tagui is an incredible tool, this is my second post and it was a pleasure to be able to help. Greetings!!!_** 👍

**Comment by RaiMan** [comment]
@pp082 
Great thanks for these additional insights. Highly appreciated. I will add the info to my wiki entry.

**Comment by MiddleAgedMutantNinjaCoder** [comment]
On MacOS 10.15.7 Catalina, I can't get the mouse access permissions (Accessibility) to work when I run the API .jar file (with a script) from the terminal.  It works OK with the IDE.

**Comment by RaiMan** [comment]
Try with Java 11 from AdoptOpenJDK and look here:
https://github.com/RaiMan/SikuliX1/wiki/Allow-SikuliX-actions-on-macOS

**Comment by RaiMan** [comment]
should work now

**Comment by titusfx** [comment]
I have the exact same problem with ubuntu

**Comment by RaiMan** [comment]
@titusfx 
what exactly? 
System? Java? SikuliX?

**Comment by dschiller** [comment]
Same Issue her with Monterey and sikulixide-2.0.5.jar. Don't see any System Settings message to aggree any Settings. Tried to manually add Java inside Systemsettings and Screen Recording. But have no Idea where to find java. Isn't it possible to lead the cirrect app like other Applications do if Permissions missing ?

**Comment by RaiMan** [comment]
@dschiller 
from 4 comments above yours:

> Try with Java 11 from AdoptOpenJDK and look here:
https://github.com/RaiMan/SikuliX1/wiki/Allow-SikuliX-actions-on-macOS

>Isn't it possible to lead the correct appto the System Settigns like other Applications do if Permissions are missing ?

Surely it is. Already had a look into it.
... but currently I have other priorities.
... but I am open to get a PR from whoever with an implementation suggestion.

**Comment by ksolariz** [comment]
Hi
I'm trying to run a sikuli script calling through a client's platform and it also has the same "mouse blocked" error I have no idea what can be done to resolve it.
![image](https://user-images.githubusercontent.com/83794987/146433774-c2760a5a-a035-42aa-aeb8-de678074f238.png)


This is my code in python that calls the sikuli script and passes some parameters to it
![image](https://user-images.githubusercontent.com/83794987/146434036-70c74ea3-56f4-4312-b371-be2fa87d93f9.png)

This is the start of my code in sikuli

![image](https://user-images.githubusercontent.com/83794987/146434898-d2add669-b26b-42be-9a80-6f93752c94e7.png)

Could someone help me solve this problem? I really have no idea what it might be.
The remote the script is trying to run on is a windows operating system with version 11 of Java.
thankful!

**Comment by RaiMan** [comment]
@ksolariz 
This thread is related to macOS.
You are on Windows: totally different situation.
Please ask such questions on Launchpad (bigger audience).

On Windows it usually helps running your stuff from an admin command line.

**Comment by ksolariz** [comment]
Oh sorry.
I thought it was the same situation because the error description is the same.
I redid the questions in other issues.

How to make this run with admin command line?

**Comment by RaiMan** [comment]
@ksolariz 
as mentioned: please post such questions on Launchpad.

final answer here:
- always test scripts first in the SikuliX IDE before using the scripts in other environments.
- a command line window on Windows can be started with admin rights

**Comment by yamoinca** [comment]
I have the same issue even after the System Preferences troubleshooting steps on M1 MBP OSX 12.3.1.
I have not tried the sqlite steps yet.

**Comment by yamoinca** [comment]
> I have the same issue even after the System Preferences troubleshooting steps on M1 MBP OSX 12.3.1. I have not tried the sqlite steps yet.

I tried by disabling sip permissions.
On reboot I removed and re-added permissions in System Preferences for java and terminal
Tried starting ide using java command line and it launched without error.

However, another app updated due to the disabling of SIP and so system needed another restart.
After restart, I got the same error again when launching using java!-(
